include/unifex/dematerialize.hpp

```diff
@@ -177,33 +177,29 @@ namespace _demat {
 
 namespace _demat_cpo {
   inline const struct _fn {
-   private:
-    template <bool>
-    struct _impl {
-      template <typename Sender>
-      auto operator()(Sender&& predecessor) const
-          noexcept(is_nothrow_tag_invocable_v<_fn, Sender>) {
-        return unifex::tag_invoke(_fn{}, (Sender&&) predecessor);
-      }
-    };
-   public:
+  private:
     template <typename Sender>
+    using _result_t =
+      typename conditional_t<
+        tag_invocable<_fn, Sender>,
+        meta_tag_invoke_result<_fn>,
+        meta_quote1<_demat::sender>>::template apply<Sender>;
+  public:
+    template(typename Sender)
+      (requires tag_invocable<_fn, Sender>)
     auto operator()(Sender&& predecessor) const
-        noexcept(is_nothrow_callable_v<
-          _impl<is_tag_invocable_v<_fn, Sender>>, Sender>) {
-      return _impl<is_tag_invocable_v<_fn, Sender>>{}((Sender&&) predecessor);
+        noexcept(is_nothrow_tag_invocable_v<_fn, Sender>)
+        -> _result_t<Sender> {
+      return unifex::tag_invoke(_fn{}, (Sender&&) predecessor);
     }
```

Comment:
after looking at a couple of these - 
  result_t and the operator() for tag_invoke look like they can be defined once in a base class and reused, right?

Reply:
Yes, but I think there is value to having it this way. It is far less surprising for users to see diagnostics pointing to the algorithm they just tried to use, rather than some common base class shared by all the algorithms.

Reply:
I assumed the base would have the algo context in the template params